packages/flutter/lib/src/material/tab_controller.dart

```diff
@@ -485,21 +494,25 @@ class _DefaultTabControllerState extends State<DefaultTabController> with Single
         newIndex = math.max(0, widget.length - 1);
         previousIndex = _controller.index;
       }
-      _controller = _controller._copyWith(
+      final TabController newController = _controller._copyWith(
         length: widget.length,
         animationDuration: widget.animationDuration,
         index: newIndex,
         previousIndex: previousIndex,
       );
+      _controller.dispose();
+      _controller = newController;
     }
 
```

Comment:
How about making it more readable:
```
TabController newController = _controller;

if (oldWidget.length != widget.length) {
  newController = newController._copyWith...
}

if (oldWidget.animationDuration != widget.animationDuration) {
  newController = newController._copyWith...
}

if (newController != _controller) {
  _controller.dispose();
  _controller = newController;
  _animationController = null;
}

```

Reply:
@polina-c, thanks a lot for your suggestion, it looks much cleaner. But if we do it in such a way, then we will lose the reference to the first new `TabController` if the `length` and `animationDuration` are changed at the same time.

Reply:
Really? Both `_copyWith` are invoked for `newController`, not for `_controller`. so, nothing is lost.

Reply:
Maybe I am missing something, but here is how I see the flow in case both `length` and `animationDuration` was updated:

```dart
// newController = TabController0
TabController newController = _controller;

if (oldWidget.length != widget.length) {
  // newController = TabController1
  newController = newController._copyWith...
}

if (oldWidget.animationDuration != widget.animationDuration) {
  // newController = TabController2
  newController = newController._copyWith...
}

if (newController != _controller) {
  _controller._animationController = null;
  // TabController0.dispose()
  _controller.dispose();
  // _controller = TabController2
  _controller = newController;

 // TabController1 is lost
}
```

Reply:
Oh, I see now.
You are right.
Thanks for explanation.

Then the only question is about `_animationController = null;`:
1. Why we need it? Old controller is not referenced and is disposed, so what's the point to null _animationController?
2. If it is needed, (1) it does not seem `_copyWith` is right method to do it and (2) there should be comment that explains why it is needed.

Reply:
Thanks for the review.

Under the hood, `_copyWith` transfers `AnimationController` from one `TabController` to another. 
```dart
  /// Creates a new [TabController] with `index`, `previousIndex`, `length`, and
  /// `animationDuration` if they are non-null.
  ///
  /// This method is used by [DefaultTabController].
  ///
  /// When [DefaultTabController.length] is updated, this method is called to
  /// create a new [TabController] without creating a new [AnimationController].
  TabController _copyWith({
  ...
  }) {
   ...
    return TabController._(
      ...
      animationController: _animationController,
      ....
    );
  }
```

In `dispose` method of the `TabController` we have disposal of the `AnimationController`. 

```dart
  @override
  void dispose() {
    _animationController?.dispose();
    _animationController = null;
    super.dispose();
  }
```

We can't hold the same `AnimationController` in both `TabController`s because it will cause double disposal in the future. So, to not dispose recently injected `AnimationController` in the old `TabController`, we need to null it. It will be disposed in the new `TabController` in the future.

This explains why I added nulling of the `_animationController` in the `_copyWith` method and added this comment:
```  
  /// This instance of [TabController] must not be used anymore and has to be
  /// disposed since [AnimationController] was injected into the new
  /// [TabController].
```

What can you suggest to improve or change?

Reply:
Great suggestion, thanks!
Pushed the new version.